lib/grandpa/grandpa_test.go

```diff
@@ -1279,3 +1280,127 @@ func TestFinalRoundGaugeMetric(t *testing.T) {
+	require.NoError(t, err)
+	require.NotNil(t, d)
+	digest := types.NewDigest()
+	_ = digest.Add(*d)
+
+	bfcBlock := &types.Block{
+		Header: types.Header{
+			ParentHash: previousHash,
+			Number:     big.NewInt(int64(bfcNumber)),
+			StateRoot:  trie.EmptyHash,
+			Digest:     digest,
+		},
+		Body: types.Body{},
+	}
+
+	bfcHash := bfcBlock.Header.Hash()
+	err = st.Block.AddBlockWithArrivalTime(bfcBlock, time.Now())
+	require.Nil(t, err)
+
+	// create fake authorities
+	fakeAuthorities := make([]*ed25519.Keypair, 0)
+	ed25519Keyring, _ := keystore.NewEd25519Keyring()
+
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Alice().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Bob().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Charlie().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Dave().(*ed25519.Keypair))
+	fakeAuthorities = append(fakeAuthorities, ed25519Keyring.Eve().(*ed25519.Keypair))
+
+	equivocatories := make(map[ed25519.PublicKeyBytes][]*types.GrandpaSignedVote)
+	prevotes := &sync.Map{}
+
+	// voting on
+	for i, v := range fakeAuthorities {
+		vote := &SignedVote{
+			AuthorityID: v.Public().(*ed25519.PublicKey).AsBytes(),
+			Signature:   [64]byte{},
+			Vote: types.GrandpaVote{
+				Hash:   bfcHash,
+				Number: uint32(bfcNumber),
+			},
+		}
+
+		prevotes.Store(i, vote)
+
+		// lets say authority 1 and 3 have equivocatory votes
+		if i == 1 || i == 3 {
+			equivocatories[v.Public().(*ed25519.PublicKey).AsBytes()] = []*types.GrandpaSignedVote{
+				{
+					Vote: types.GrandpaVote{
+						Hash:   common.Hash{},
+						Number: 0,
+					},
+					Signature:   [64]byte{},
+					AuthorityID: v.Public().(*ed25519.PublicKey).AsBytes(),
+				},
+			}
+		}
+	}
+
+	gs.pvEquivocations = equivocatories
+	gs.prevotes = prevotes
+
+	justifications, err := gs.createJustification(bfcHash, prevote)
+	require.NoError(t, err)
+	require.Len(t, justifications, len(fakeAuthorities)+len(equivocatories))
+}
+
+// addBlocksToState test helps adding previous blocks
+func addBlocksToState(t *testing.T, blockState *state.BlockState, depth int) {
+	t.Helper()
+
+	previousHash := blockState.BestBlockHash()
+	arrivalTime := time.Now()
+
+	rt, err := blockState.GetRuntime(nil)
+	require.NoError(t, err)
+
+	head, err := blockState.BestBlockHeader()
+	require.NoError(t, err)
+
+	startNum := int(head.Number.Int64())
+
+	for i := startNum + 1; i <= depth; i++ {
+		d, err := types.NewBabePrimaryPreDigest(0, uint64(i), [32]byte{}, [64]byte{}).ToPreRuntimeDigest()
+		require.NoError(t, err)
+		require.NotNil(t, d)
+		digest := types.NewDigest()
+		_ = digest.Add(*d)
+
+		block := &types.Block{
+			Header: types.Header{
+				ParentHash: previousHash,
+				Number:     big.NewInt(int64(i)),
+				StateRoot:  trie.EmptyHash,
+				Digest:     digest,
+			},
+			Body: types.Body{},
```

Comment:
nit

```suggestion
```

Reply:
this is required otherwise I will got `empty block body`, this is why `Body` is `[]byte` so if I don't pass the default value will be nil afaik

Reply:
shouldn't the error `empty block body` be returned even for a zero length body? :thinking:

Reply:
@EclesioMeloJunior just a small reminder :wink:

Reply:
@qdm12, sorry for the delay, the right error is `block body is nil`, @noot in the function `AddBlockWithArrivalTime`, can I change the check `block.Body == nil` to `len(block.Body) == 0`?

Reply:
could you not check both?